internal/internal_workflow.go

```diff
@@ -1356,60 +1355,52 @@ func (h *queryHandler) execute(input []byte) (result []byte, err error) {
 }
 
 // Add increments the WaitGroup counter indicating that
-// a coroutine has been added to the WaitGroup and will be
-// waited on.
+// a coroutine has been added to the WaitGroup.
 //
-// param ctx Context -> workflow context
-//
-// param delta int -> the value to increment the WaitGroup counter by
-//
-// returns WaitGroup
-func (wg *waitGroupImpl) Add(ctx Context, delta int) WaitGroup {
-	if delta < 0 {
-		panic("cannot add negative value %d to WaitGroup")
-	}
-
-	for i := 0; i < delta; i++ {
-		f, _ := NewFuture(ctx)
-		wg.selector = wg.selector.AddFuture(f, func(future Future) {
-			wg.mu.Lock()
-			defer wg.mu.Unlock()
-			wg.n = wg.n - 1
-			if wg.n == 0 {
-				wg.ready = true
-			}
-		}).(*selectorImpl)
-	}
-
-	// increment
-	wg.mu.Lock()
-	defer wg.mu.Unlock()
+// param delta int32 -> the value to increment the WaitGroup counter by
+func (wg *waitGroupImpl) Add(delta int32) {
+	state := wg.waiting
 	wg.n = wg.n + delta
-
-	return wg
+	if wg.n < 0 {
+		panic("negative WaitGroup counter")
+	}
+	if (wg.waiting && delta > 0) && (wg.n == delta) {
+		panic("WaitGroup misuse: Add called concurrently with Wait")
+	}
+	if (wg.n > 0) || (!wg.waiting) {
+		return
+	}
+	if wg.waiting != state {
```

Comment:
This check is not necessary as preemption in Cadence workflow is voluntary. So it is not possible that this method is preempted while running.

Reply:
Great I'll remove the last check.